kvdb/cache.go

```diff
@@ -0,0 +1,930 @@
+	c.pending.PushBack(&pendingAdd{parent, newChild})
+}
+
+func (c *Cache) pendingUpdate(parent *cachedItem, oldValue string) {
+	c.pending.PushBack(&pendingUpdate{parent, oldValue})
+}
+
+func (c *Cache) pendingDelete(parent *cachedItem, oldChild *cachedItem) {
+	c.pending.PushBack(&pendingDelete{parent, oldChild})
+}
+
+type traversalHelper struct {
+	bucket walletdb.ReadBucket
+	root   *cachedItem
+}
+
+// AddTopLevelBucket recursively reads the passed top-level bucket and adds
+// all content below it to the cache.
+func (c *Cache) AddTopLevelBucket(key []byte) error {
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	var root *cachedItem
+
+	err := View(c.backend, func(tx RTx) error {
+		var queue []traversalHelper
+
+		currBucket := tx.ReadBucket(key)
+		if currBucket == nil {
+			return nil
+		}
+
+		root = &cachedItem{
+			key:    string(key),
+			bucket: newCacheBucket(),
+		}
+
+		currRoot := root
+
+		for {
+			currBucket.ForEach(func(k, v []byte) error {
+				var item *cachedItem
+				if v == nil {
+					item = &cachedItem{
+						key:    string(k),
+						bucket: newCacheBucket(),
+					}
+
+					bucket := currBucket.NestedReadBucket(k)
+					queue = append(
+						queue,
+						traversalHelper{
+							bucket: bucket,
+							root:   item,
+						},
+					)
+				} else {
+					item = &cachedItem{
+						key:   string(k),
+						value: string(v),
+					}
+
+				}
+
+				currRoot.bucket.put(item)
+				return nil
+			})
+
+			if len(queue) == 0 {
+				break
+			}
+
+			currBucket = queue[0].bucket
+			currRoot = queue[0].root
+			queue = queue[1:]
+		}
+
+		return nil
+	}, func() {})
+
+	if err != nil {
+		return err
+	}
+
+	if root != nil {
+		c.root.bucket.put(root)
+	}
+
+	return nil
+}
+
+// Wipe wipes the cache state.
+func (c *Cache) Wipe() {
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	c.root = &cachedItem{
+		bucket: newCacheBucket(),
+	}
+}
+
+func (c *Cache) BeginReadTx() (walletdb.ReadTx, error) {
+	c.mx.RLock()
+
+	dbTx, err := c.backend.BeginReadTx()
+	if err != nil {
+		c.mx.RUnlock()
+		return nil, err
+	}
+
+	return newCacheReadTx(c, dbTx), nil
+}
+
+func (c *Cache) BeginReadWriteTx() (walletdb.ReadWriteTx, error) {
+	c.mx.Lock()
+
+	dbTx, err := c.backend.BeginReadWriteTx()
+	if err != nil {
+		c.mx.Unlock()
+		return nil, err
+	}
+
+	c.currRwTx = newCacheReadWriteTx(c, dbTx)
+	return c.currRwTx, nil
+}
+
+func (c *Cache) Copy(w io.Writer) error {
+	return fmt.Errorf("unavailable")
+}
+
+func (c *Cache) Close() error {
+	err := c.backend.Close()
+	c.Wipe()
+	return err
+}
+
+func (c *Cache) PrintStats() string {
+	return "<unavailable>"
+}
+
+func (c *Cache) View(f func(tx walletdb.ReadTx) error, reset func()) error {
+	tx, err := c.BeginReadTx()
+	if err != nil {
+		return err
+	}
+
+	reset()
+
+	err = f(tx)
+	rollbackErr := tx.Rollback()
+
+	if err != nil {
+		return err
+	}
+
+	return rollbackErr
+}
+
+func (c *Cache) Update(f func(tx walletdb.ReadWriteTx) error,
```

Comment:
Is this compatible w/ the retry logic of etcd?

Reply:
So the idea is that if the `kvdb.Cache` is used, it is the single writer of the added top level buckets so there won't be any conflicts. If the user edits the DB while LND is working, this assumption breaks but I think that use case is out of scope.